File: 49.rb
# Given an array of strings, group anagrams together.

# For example, given: ["eat", "tea", "tan", "ate", "nat", "bat"],
# Return:

# [
#   ["ate", "eat","tea"],
#   ["nat","tan"],
#   ["bat"]
# ]

# Note: All inputs will be in lower-case.

# @param {String[]} strs
# @return {String[][]}
def group_anagrams1(strs)
  anagrams = []
  hashed_arr = strs.map { |str| calculate_char_occur_times str }

  hashed_arr.each_with_index do |hsh, idx|
    grouped = false
    anagrams.each do |group|
n      if hashed_arr[group.first] == hsh
        group << idx
        grouped = true
      end

    end
    anagrams << [idx] unless grouped
  end
  anagrams.map { |group_of_idx| group_of_idx.map { |idx| strs[idx] } }
end

# decide if two words are anagrams
# iterate through all characters, store each chracter occur
# times into a hash.
# compare hash to decide if it's anagram

def calculate_char_occur_times(str)
  hsh = Hash.new { |h, k| h[k] = 0 }
  str.each_char { |c| hsh[c] += 1 }
  hsh
end

def group_anagrams2(strs)
  anagrams = []
  size = strs.size
  while !size.zero?
    target = strs.first
    group = [target]
    strs[1, strs.size - 1].each do |str|
      group << str if anagram?(target, str)
    end
    anagrams << group
    strs -= group
    size = strs.size
  end
  anagrams
end

def anagram?(str1, str2)
  return false if str1.size != str2.size
  str1 == str2
end

# both methods of group_anagrams1 and group_anagrams2 exceeds time limit
# the problem here is that every word needs a calculation of anagram?
# but this is not necessary, just convert word into sorted form and put into array

def group_anagrams(strs)
  anagrams = Hash.new([])
  strs.each_with_index do |str, idx|
    anagrams[str.chars.sort.join] += [idx]
  end
  anagrams.reduce([]) { |memo, obj| memo << obj[1].map { |idx| strs[idx] } }
end

p group_anagrams(["eat","tea","tan","ate","nat","bat"])

File: 49.go
package main

// Given an array of strings, group anagrams together.
//
// Example:
//
// Input: ["eat", "tea", "tan", "ate", "nat", "bat"],
// Output:
// [
//   ["ate","eat","tea"],
//   ["nat","tan"],
//   ["bat"]
// ]
//
// Note:
//
//     All inputs will be in lowercase.
//     The order of your output does not matter.

// array can be used in map
type table [26]int

func groupAnagrams(strs []string) [][]string {
	counter := make(map[table][]string)

	for _, str := range strs {
		var c table
		for i := range str {
			c[str[i]-'a']++
		}
		counter[c] = append(counter[c], str)
	}

	ans := make([][]string, 0)

	for _, v := range counter {
		ans = append(ans, v)
	}

	return ans
}

func groupAnagrams1(strs []string) [][]string {
	result := make([][]string, 0)
	signatures := make(map[string][]string)

	var j int
	tmp := make([]int, 26)
	for i := range strs {
		reset(&tmp)

		for j = range strs[i] {
			tmp[strs[i][j]-'a']++
		}

		key := mapKey(tmp)
		if _, ok := signatures[key]; ok {
			signatures[key] = append(signatures[key], strs[i])
		} else {
			signatures[key] = []string{strs[i]}
		}
	}

	for _, v := range signatures {
		result = append(result, v)
	}

	return result
}

func reset(tmp *[]int) {
	for i := range *tmp {
		(*tmp)[i] = 0
	}
}

// tea -> a1e1t1
func mapKey(data []int) string {
	var b []byte

	for i := range data {
		if data[i] > 0 {
			b = append(b, byte('a'+i))
			b = append(b, byte(data[i])-byte(0))
		}
	}

	return string(b)
}

//	Notes
//	1.	too slow, remove some memory allocation

//	2.	still too slow, I think problems comes from too many groups to search,
//		I can use map for faster find, with custom key structure tea -> a1e1t1

//	3.	when grouping, it's a hidden criteria that dedup is needed

//	4.	use map for faster finding, but be aware that to convert char # into
//		byte, it needs to deduct byte(0), otherwise it might cause duplicates

//	5.	from reference https://leetcode.com/problems/group-anagrams/discuss/19176/Share-my-short-JAVA-solution

//		it's also possible to convert those char bucket into string

//	6.	inspired from https://leetcode.com/problems/group-anagrams/discuss/19233/O(M-*-N)-algorithm-using-hash-without-sort()

//		author uses [26]primes to get encoded key (a * 2, b * 3, c * 5, etc.), thus
//		sort is not used

//	7.	inspired from sample code, after counting all char frequency, convert it
//		to lexicographical string, it can be used as key because anagrams with
//		same lexicographical string (sort could also work)
